webcam-capture-drivers/driver-ffmpeg-cli/src/main/java/com/github/sarxos/webcam/ds/ffmpegcli/FFmpegCliDevice.java

```diff
@@ -1,260 +1,218 @@
-			"-s", getResolutionString(),   // frame dimension
-			"-i", vfile.getAbsolutePath(), // input file 
-			"-vcodec", "copy",             // video codec to be used
-			pipe.getAbsolutePath(),        // output file
-		};
-		//@formatter:on
-
-		if (LOG.isDebugEnabled()) {
-			StringBuilder sb = new StringBuilder();
-			for (String c : cmd) {
-				sb.append(c).append(' ');
-			}
-			LOG.debug("Executing command: {}", sb.toString());
-		}
-
-		try {
-			RT.exec(new String[] { "mkfifo", pipe.getAbsolutePath() }).waitFor();
-			process = RT.exec(cmd);
-			dis = new DataInputStream(new FileInputStream(pipe));
-		} catch (FileNotFoundException e) {
-			throw new RuntimeException(e);
-		} catch (IOException e) {
-			throw new RuntimeException(e);
-		} catch (InterruptedException e) {
-			throw new RuntimeException(e);
-		}
-	}
-
-	@Override
-	public synchronized void close() {
-
-		if (!open.compareAndSet(true, false)) {
-			return;
-		}
-
-		try {
-			dis.close();
-		} catch (IOException e) {
-			throw new RuntimeException(e);
-		}
-
-		process.destroy();
-
-		try {
-			process.waitFor();
-		} catch (InterruptedException e) {
-			throw new RuntimeException(e);
-		}
-
-		if (!pipe.delete()) {
-			pipe.deleteOnExit();
-		}
-	}
-
-	@Override
-	public void dispose() {
-		if (disposed.compareAndSet(false, true) && open.get()) {
-			close();
-		}
-	}
-
-	@Override
-	public boolean isOpen() {
-		return open.get();
-	}
-
-	@Override
-	public String toString() {
-		return "video device " + name;
-	}
+    private static final Logger LOG = LoggerFactory.getLogger(FFmpegCliDevice.class);
+
+    private Process process = null;
+
+    private File vfile = null;
+    private String name = null;
+    private Dimension[] resolutions = null;
+    private Dimension resolution = null;
+
+    private AtomicBoolean open = new AtomicBoolean(false);
+    private AtomicBoolean disposed = new AtomicBoolean(false);
+
+    BufferedImage image;
+
+    protected FFmpegCliDevice(File vfile, String vinfo) {
+        String[] parts = vinfo.split(" : ");
+
+        this.vfile = vfile;
+        this.name = vfile.getAbsolutePath();
+        this.resolutions = readResolutions(parts[3].trim());
+
+        Dimension resolution = getResolution();
+        image = new BufferedImage(resolution.width, resolution.height, BufferedImage.TYPE_3BYTE_BGR);
+    }
+
+    public void startProcess() throws IOException {
+        ProcessBuilder builder = new ProcessBuilder(buildCommand());
+        builder.redirectErrorStream(true); // so we can ignore the error stream
+
+        Process process = builder.start();
+        InputStream out = process.getInputStream();
+        OutputStream in = process.getOutputStream();
+
+        int SIZE = resolution.width * resolution.height * 3;
+        int CHUNK_SIZE = SIZE / 20;
+
+        int cursor = 0;
+
+        byte[] chunk = new byte[CHUNK_SIZE];
+        ByteBuffer buffer = ByteBuffer.allocate(SIZE);
+
+        while (isAlive(process)) {
```

Comment:
I tested this code and I'm always getting black image. What is more important - after I close window, the camera remained open (LED was still powered on) and I was unable to open it again. Even when I tried from console I got ioctl error:

```
$ ffmpeg -f v4l2 -framerate 30 -s 640x480 -i /dev/video0 -pix_fmt bgr24 -vsync vfr -vcodec rawvideo output.mkv
ffmpeg version 3.3.3 Copyright (c) 2000-2017 the FFmpeg developers
  built with gcc 4.8 (Ubuntu 4.8.4-2ubuntu1~14.04.3)
  configuration: --extra-libs=-ldl --prefix=/opt/ffmpeg --mandir=/usr/share/man --enable-avresample --disable-debug --enable-nonfree --enable-gpl --enable-version3 --enable-libopencore-amrnb --enable-libopencore-amrwb --disable-decoder=amrnb --disable-decoder=amrwb --enable-libpulse --enable-libfreetype --enable-gnutls --disable-ffserver --enable-libx264 --enable-libx265 --enable-libfdk-aac --enable-libvorbis --enable-libtheora --enable-libmp3lame --enable-libopus --enable-libvpx --enable-libspeex --enable-libass --enable-avisynth --enable-libsoxr --enable-libxvid --enable-libvidstab --enable-libwavpack --enable-nvenc --enable-libzimg
  libavutil      55. 58.100 / 55. 58.100
  libavcodec     57. 89.100 / 57. 89.100
  libavformat    57. 71.100 / 57. 71.100
  libavdevice    57.  6.100 / 57.  6.100
  libavfilter     6. 82.100 /  6. 82.100
  libavresample   3.  5.  0 /  3.  5.  0
  libswscale      4.  6.100 /  4.  6.100
  libswresample   2.  7.100 /  2.  7.100
  libpostproc    54.  5.100 / 54.  5.100
[video4linux2,v4l2 @ 0x3db9e20] ioctl(VIDIOC_STREAMON): Invalid argument
/dev/video0: Invalid argument
```

It was probably caused by the incorrectly closed sub-process since when I restarted my Linux machine everything went back to normal (till I executed your code again few times effectively causing next ioctl error).

It's locked in some weird state. Other drivers stopped working as well:

![image](https://user-images.githubusercontent.com/472658/35187739-dd5d0474-fe28-11e7-9c9c-261f411bb775.png)

```plain
Caused by: com.github.sarxos.webcam.WebcamException: Cannot start native grabber!
	at com.github.sarxos.webcam.ds.buildin.WebcamDefaultDevice.open(WebcamDefaultDevice.java:334) ~[classes/:na]
	at com.github.sarxos.webcam.ds.cgt.WebcamOpenTask.handle(WebcamOpenTask.java:38) ~[classes/:na]
	at com.github.sarxos.webcam.WebcamProcessor$AtomicProcessor.run(WebcamProcessor.java:81) ~[classes/:na]
	... 3 common frames omitted
```

Reply:
Sorry for that, my bad.